ddtrace/tracer/abandoned_spans.go

```diff
@@ -0,0 +1,190 @@
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+type AbandonedList struct {
+	head *spansList
+	tail *spansList
+}
+
+type spansList struct {
+	head *spanNode
+	tail *spanNode
+	Next *spansList
+}
+
+type spanNode struct {
+	Element *span
+	Next    *spanNode
+}
+
+func (a *AbandonedList) String() string {
+	var sb strings.Builder
+	for e := a.head; e != nil; e = e.Next {
+		fmt.Fprintf(&sb, "%v", e.String())
+	}
+	return sb.String()
+}
+
+func (s *spansList) String() string {
+	var sb strings.Builder
+	for e := s.head; e != nil; e = e.Next {
+		fmt.Fprintf(&sb, "%v", e.String())
+	}
+	return sb.String()
+}
+
+func (s *spanNode) String() string {
+	sp := s.Element
+	if sp == nil {
+		return "[],"
+	}
+	return fmt.Sprintf("[Span Name: %v, Span ID: %v, Trace ID: %v],", sp.Name, sp.SpanID, sp.TraceID)
+}
+
+func (a *AbandonedList) Extend() {
+	n := &spansList{}
+	if a.head == nil {
+		a.head = n
+		a.tail = n
+		return
+	}
+	a.tail.Next = n
+	a.tail = n
+}
+
+func (a *AbandonedList) RemoveTail() {
+	n := a.head
+	if n == nil || n.Next == nil {
+		a.head = nil
+		return
+	}
+	for n.Next.Next != nil {
+		n = n.Next
+	}
+	n.Next = nil
+	a.tail = n
+}
+
+func (a *AbandonedList) RemoveBucket(s *spansList) {
+	if s.Next != nil {
+		n := s.Next
+		s.head = n.head
+		s.Next = n.Next
+		return
+	}
+	a.RemoveTail()
+}
+
+func (s *spansList) Append(e *span) {
+	n := &spanNode{Element: e}
+	if s.head == nil {
+		s.head = n
+		s.tail = n
+		return
+	}
+	s.tail.Next = n
+	s.tail = n
+}
+
+func (s *spansList) Remove(e *span) {
+	if s.head == nil {
+		return
+	}
+	if s.head.Element == e {
+		s.head = s.head.Next
+		return
+	}
+
+	n := s.head
+	for n.Next != nil {
+		if n.Next.Element == e {
+			n.Next = n.Next.Next
+			if n.Next == nil {
+				s.tail = n
+			}
+			return
+		}
+		n = n.Next
+	}
+}
+
+var tickerInterval = time.Minute
+
+// reportAbandonedSpans periodically finds and reports potentially
+// abandoned spans that are older than the given interval
+func (t *tracer) reportAbandonedSpans(interval time.Duration) {
+	tick := time.NewTicker(tickerInterval)
+	defer tick.Stop()
+	for {
+		select {
+		case <-tick.C:
+			for b := t.abandonedSpans.head; b != nil; b = b.Next {
+				e := b.head
+				if e == nil || e.Element == nil {
+					continue
+				}
+				if now()-e.Element.Start < interval.Nanoseconds() {
+					continue
+				}
+				for e != nil {
+					sp := e.Element
+					life := now() - sp.Start
+					if life >= interval.Nanoseconds() {
+						log.Warn("Trace %v waiting on span %v", sp.Context().TraceID(), sp.Context().SpanID())
+						e = e.Next
+					} else {
+						break
+					}
+				}
+			}
+		case s := <-t.cIn:
+			e := t.abandonedSpans.head
+			for e != nil {
+				sp := e.head
+				if sp == nil || sp.Element == nil {
+					e = &spansList{}
+					e.Append(s)
+					break
+				}
+				if s.Start-sp.Element.Start <= interval.Nanoseconds() {
+					e.Append(s)
+					break
+				}
+				e = e.Next
+			}
+			if e != nil {
+				break
+			}
```

Comment:
Sorry I'm not following what `e` is used for here, and what this code is doing at first glance? Can you clarify why we can't just make a single function call to add this element to the abandoned spans list, and make any extra logic just a method on the list?

Reply:
Fair point; I might've chosen better variable names here :') The `abandonedSpans` list is bucketed, so instead of adding all new spans to the list, I try to determine which "bucket" (`e`) the span should belong to. This is why I have the check for the start time of the new span and existing spans in the list (I want to put spans that started around the same time as each other into the same bucket, so that when looking for abandoned spans I could potentially exit early). 

Overall, I think I could clear up the code a bit, even if it just means using slightly longer variable names. I was trying to keep them short in the spirit of using Golang 😆 What do you think?

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#variable-names describes this nicely:

> The basic rule: the further from its declaration that a name is used, the more descriptive the name must be.

So it's a judgement call.
If one of those variables is the bucket it should go into, then `bucket` would be a great variable name here. I'm not sure what `sp` here represents

Reply:
Good point :) I'll clear up the names! Thanks 🎉